crates/rpc/rpc/src/eth/gas_oracle.rs

```diff
@@ -0,0 +1,264 @@
+
+    /// The default gas price to use if there are no blocks to use
+    pub default: Option<U256>,
+
+    /// The maximum gas price to use for the estimate
+    pub max_price: Option<U256>,
+
+    /// The minimum gas price, under which the sample will be ignored
+    pub ignore_price: Option<U256>,
+}
+
+impl Default for GasPriceOracleConfig {
+    fn default() -> Self {
+        GasPriceOracleConfig {
+            blocks: 20,
+            percentile: 60,
+            max_header_history: 1024,
+            max_block_history: 1024,
+            default: None,
+            max_price: Some(DEFAULT_MAX_PRICE),
+            ignore_price: Some(DEFAULT_IGNORE_PRICE),
+        }
+    }
+}
+
+/// Calculates a gas price depending on recent blocks.
+#[derive(Debug)]
+pub struct GasPriceOracle<Client> {
+    /// The type used to subscribe to block events and get block info
+    client: Client,
+    /// The cache for blocks
+    cache: EthStateCache,
+    /// The config for the oracle
+    oracle_config: GasPriceOracleConfig,
+    /// The latest calculated price and its block hash
+    last_price: Mutex<GasPriceOracleResult>,
+}
+
+impl<Client> GasPriceOracle<Client>
+where
+    Client: BlockProviderIdExt + 'static,
+{
+    /// Creates and returns the [GasPriceOracle].
+    pub fn new(
+        client: Client,
+        mut oracle_config: GasPriceOracleConfig,
+        cache: EthStateCache,
+    ) -> Self {
+        // sanitize the perentile to be less than 100
+        if oracle_config.percentile > 100 {
+            warn!(prev_percentile=?oracle_config.percentile, "Invalid configured gas price percentile, using 100 instead");
+            oracle_config.percentile = 100;
+        }
+
+        Self { client, oracle_config, last_price: Default::default(), cache }
+    }
+
+    /// Suggests a gas price estimate based on recent blocks, using the configured percentile.
+    pub async fn suggest_tip_cap(&self) -> EthResult<U256> {
+        let header_hash = self
+            .client
+            .block_hash_for_id(BlockId::Number(BlockNumberOrTag::Latest))?
+            .expect("a latest header should always exist");
+        let header = self
+            .client
+            .header_by_hash_or_number(BlockHashOrNumber::Hash(header_hash))?
+            .ok_or(GeneralError::Provider(ProviderError::BlockHash { block_hash: header_hash }))?;
+
+        let mut last_price = self.last_price.lock().await;
+
+        // if we have stored a last price, then we check whether or not it was for the same head
+        if last_price.block_hash == header_hash {
+            return Ok(last_price.price)
+        }
+
+        // if all responses are empty, then we can return a maximum of 2*check_block blocks' worth
+        // of prices
+        //
+        // we only return more than check_block blocks' worth of prices if one or more return empty
+        // transactions
+        let mut current_block = header.number;
+        let mut results = Vec::new();
```

Comment:
can we pre allocate here, or doesn't really matter?

Reply:
Not sure how much we should pre allocate, because we don't know how many values `get_block_values` will return, or exactly how many times we will call it